Basic integration tests for pool registrations and retirements.

jonathanknowles opened the pull request:
# Issue Number

#1819 

# Overview

This PR provides a first step towards a full suite of integration tests for pool registrations and retirements.

- [x] Adds function `issuePoolRetirementCert` capable of creating Shelley pool retirement certificates.
- [x] Adjusts the initialization of the test pool cluster so that a subset of the pools are configured to retire after different time intervals.
- [x] Adds a basic integration test to verify that the `ListStakePools` API operation eventually shows the correct retirement epochs for the test pools.

# Comments

Future PRs will:

- Simulate the effect of re-registering a pool after a retirement certificate has already been published.
- Verify that it is not possible to join a stake pool that has already retired (see #1913).

jonathanknowles committed:
Rename `issuePoolCert` to `issuePoolRegistrationCert`.

Justification: later on, we will need to distinguish between different
types of pool certificate.

jonathanknowles committed:
Add function `issuePoolRetirementCert`.

jonathanknowles committed:
Reformat code to comply with coding standards.

jonathanknowles committed:
Add optional retirement epoch argument to `setupStakePoolData`.

jonathanknowles committed:
Add optional retirement cert to the initial pool registration transaction.

jonathanknowles committed:
Add optional retirement epoch argument to `withStakePool`.

jonathanknowles committed:
Specify a very large upper bound for the Emax protocol parameter.

(For testing purposes only.)

jonathanknowles committed:
Pass list of `PoolConfig` to `withCluster` instead of simple integer.

jonathanknowles committed:
Verify that the test stake pools are not due to retire.

jonathanknowles committed:
Simulate real retirements in the stake pool tests.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -593,6 +596,23 @@ spec = do
+                    , expectListField 0 #retirement
+                        (`shouldBe` Nothing)
+                    , expectListField 1 #retirement
+                        (`shouldSatisfy` retirementEpochIs 10)
```
This test passed when I run it alone... but it was not passing when I was running it with a test suite. 

I started to monitor the cluster while the tests were running and my observation is that this pool that is scheduled to retire in epoch 10 disappears from the `list stake pools` at the beginning of epoch 14... :thinking:  so, we're starting with 3 pools being registered in "setup" but since epoch 14 wallet actually shows only 2 pools!

jonathanknowles replied:
@piotr-iohk wrote:

> This test passed when I run it alone... but it was not passing when I was running it with a test suite.
> 
> I started to monitor the cluster while the tests were running and my observation is that this pool that is scheduled to retire in epoch 10 disappears from the `list stake pools` at the beginning of epoch 14... thinking so, we're starting with 3 pools being registered in "setup" but since epoch 14 wallet actually shows only 2 pools!

Thanks for analysing this!

I'm a little surprised that it just disappears from the list, as we don't (yet) actively perform any garbage collection of pools after they retire. If I were to take a guess, the cause would be related to the way we are merging data in order to produce the list of stake pools. I'll put this next on my list for immediate investigation.

For now, I've uploaded a patch (e80b7f5b46bb8d6e94147b6f26c2c86925d092c8) which prevents any stake pools from prematurely retiring during a run of the integration test suite, which should fix this PR.

Later on, in a further PR, I'll add more integration tests to check the behaviour of pools after they retire.

jonathanknowles committed:
Ensure that no test stake pools retire prematurely.

For now, we actually want all test stake pools to remain alive for the
full duration of an integration test suite run, hence we configure our
test pools to have retirement dates that are well into the future.

In future, we will add more tests to simulate what happens after a stake
pool has retired.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1931#discussion_r457417183

jonathanknowles committed:
Use "notice" instead of "disclaimer".

Justification: this is an informational notice to assist with debugging.

(A disclaimer would be appropriate if it were necessary to limit the
scope of legal liability in some way, but that doesn't appear to be
necessary here.)

jonathanknowles committed:
Report the last error detected when `eventually` fails.
